### PR TITLE
Redirect archive [PD-1001]

### PR DESCRIPTION
**Pull request by almathew**
All tests pass.

### CODE REVIEW

File: redirect/tests/test_models.py [hunk 1]
```diff
@@ -1,36 +1,54 @@
...
+class BaseRedirectTests(TestCase):
+    def test_get_any(self):
+        redirect = factories.RedirectFactory()
+        archived_redirect = factories.RedirectArchiveFactory()
+
+        models.Redirect.objects.get_any(guid=redirect.guid)
+        models.Redirect.objects.get_any(guid=archived_redirect.guid)
+
+        self.assertRaises(ObjectDoesNotExist,
+                          models.Redirect.objects.get, guid='a')
```
Comment (tdphillips) [inline redirect/tests/test_models.py]: Should this also be `.get_any`?

Comment (almathew) [inline redirect/tests/test_models.py]: Fixed.

**Comment by almathew** [comment]
I've added the routers, but until the new databases are in place I can't actually update the deploy/settings*.py files.

This is the first time we're doing this, so feedback/opinions would be valuable.


Remaining work:
* Get new databases in place
* Figure out a way to compare guids across databases

File: deploy/settings.staging.py [hunk 2]
```diff
@@ -12,6 +12,14 @@
         'PASSWORD': PROD_DB_PASSWD,
         'HOST': 'db-redirectstaging.c9shuxvtcmer.us-east-1.rds.amazonaws.com',
         'PORT': '3306',
+    },
+    'archive': {
+        'ENGINE': 'django.db.backends.mysql',
+        'NAME': 'redirect',
+        'USER': 'de_dbuser',
+        'PASSWORD': PROD_DB_PASSWD,
+        'HOST': 'db-redirectarchivestaging.c9shuxvtcmer.us-east-1.rds.amazonaws.com',
```
Comment (almathew) [inline deploy/settings.staging.py]: This database now exists.

File: redirect/tasks.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    """
+    Adds an iteratble of redirects to to_model.
+
+    :param to_model: A subclass of BaseRedirect (either
+                     Redirect or RedirectArchive) that will
+                     be getting the copies of the redirects.
+    :param redirects: An iterable of redirect objects that should
+                      be added to the to_model table.
+    """
+    new_redirects = []
```
Comment (aspidites) [inline redirect/tasks.py]: List comprehension would be more efficient here. 

I expect `to_model.objects.bulk_create(copy_redirect(to_model, redirect) for redirect in redirects)` would also work, but wouldn't be offended if you didn't use a generator expression.

File: redirect/tasks.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                  redirect should be made for.
+    :param existing_redirect: An existing redirect instance that the new
+                              redirect should copy from.
+
+    :return: An instace of model with data copied from existing_redirect.
+             This instance has not yet been written to the table
+             for model.
+
+    """
+    return model(guid=existing_redirect.guid,
```
Comment (aspidites) [inline redirect/tasks.py]: Is this doing anything that this doesn't?:
```python
def copy_redirect(redirect):
    redirect.pk = None
    redirect.save()
    return redirect
````

Comment (tdphillips) [inline redirect/tasks.py]: PK is the guid, so setting it to None is not desired.

Comment (almathew) [inline redirect/tasks.py]: Yes. It's used for changing the model it points to. Redirect and RedirectArchive have the same schema, so you can use all the same variables, but you can't use the same object.

File: redirect/routers.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+from redirect.models import RedirectArchive
+
+
+class ArchiveRouter(object):
```
Comment (aspidites) [inline redirect/routers.py]: Where/when is this used? Quick ctrl-f came up empty, so I'm sure I'm missing something obvious. I know what it's supposed to do generally, but would love to see an example of it actually being used. I suspect that these can be protected methods (yeah, I know this is python, bare with me) on the mixin type you created.

Comment (almathew) [inline redirect/routers.py]: It's in default_settings. It's used on every query to the db.

File: redirect/tasks.py (hunk 4 above)
Comment (aspidites) [inline redirect/tasks.py]: @almathew Ah, that definitely makes sense. :+1:  for the original then.

**Pull request merged**